README.md
Introduction
------------

*flAccurateRip* is a command line tool to verify the accuracy of FLAC files
ripped from CD, using the information contained in the AccurateRip™ database
about other rips of the same CD.

Usage
-----

Suppose you ripped a CD into `track01.flac`, `track02.flac`, ...,
`trackNN.flac`. Open a shell in the directory containing the FLAC files, and
run:

    $ flaccuraterip track01.flac track02.flac ... trackNN.flac

If a pressing of this CD is present in the AccurateRip™ database,
`flaccuraterip` will output the ripping accuracy for each track. If the CD was
ripped without setting the drive offset, you should use the flag
`--with-sample-offset=N`, where N is the offset indicated [here][offsets] for
your drive.

  [offsets]:  http://accuraterip.com/driveoffsets.htm
              "http://accuraterip.com/driveoffsets.htm"


CHANGELOG.md
Version 0.3.9 (2020-06-10)
--------------------------

- Allow dependency on `optparse-applicative` 0.15.

Version 0.3.8 (2017-11-18)
--------------------------

- Allow dependency on `optparse-applicative` 0.14.
- Require `process` >= 1.5.

Version 0.3.7 (2016-11-19)
--------------------------

- Allow dependency on `optparse-applicative` 0.13.

Version 0.3.6 (2016-08-28)
--------------------------

- Allow dependency on `binary` 0.8.
- Require `process` 1.4.

Version 0.3.5 (2016-01-20)
--------------------------

- Allow dependency on `HTTP` 4000.3.
- Allow dependency on `optparse-applicative` 0.12.

Version 0.3.4 (2015-04-06)
--------------------------

- Allow dependency on `deepseq` 1.4.

Version 0.3.3 (2014-10-10)
--------------------------

- Allow dependency on `optparse-applicative` 0.11.

Version 0.3.2 (2014-09-24)
--------------------------

- Require `optparse-applicative` 0.10.
- Use autogenerated `Paths_flaccuraterip` module to import version in `Main`.

Version 0.3.1 (2014-05-25)
--------------------------

- Allow dependency on `optparse-applicative` 0.9.

Version 0.3.0 (2014-04-27)
--------------------------

- Replace `cmdargs` with `optparse-applicative` for parsing command line.
- Improve description and help message.

Version 0.2.1 (2014-03-21)
--------------------------

- Clean up code.

Version 0.2 (2012-11-08)
------------------------

- New option `--show-database-entry` to show the AccurateRip database entry of
  the rip.
- Improve description.
- Minor code changes.

Version 0.1 (2012-11-03)
------------------------

- Initial release.


CD/CDDB.hs
{-# LANGUAGE GeneralizedNewtypeDeriving #-}

module CD.CDDB where

import           Control.DeepSeq
import           Data.Binary
import           Data.Binary.Get
import           Data.Binary.Put
import           Data.Bits
import           Data.Char

newtype DiscID = DiscID Word32
    deriving (Eq, NFData)

instance Binary DiscID where
    put (DiscID w) = putWord32le w
    get = fmap DiscID getWord32le

instance Show DiscID where
    show (DiscID w) =
        [ intToDigit (fromIntegral (shiftR w 28 .&. 15))
        , intToDigit (fromIntegral (shiftR w 24 .&. 15))
        , intToDigit (fromIntegral (shiftR w 20 .&. 15))
        , intToDigit (fromIntegral (shiftR w 16 .&. 15))
        , intToDigit (fromIntegral (shiftR w 12 .&. 15))
        , intToDigit (fromIntegral (shiftR w 8 .&. 15))
        , intToDigit (fromIntegral (shiftR w 4 .&. 15))
        , intToDigit (fromIntegral (w .&. 15))
        ]

-- |Calculate CDDB DiscID: the result is XXYYYYZZ where XX is the number of tracks, YYYY is the total length of the CD is seconds, and ZZ is ...
cddbDiscId :: [Int] -> DiscID
cddbDiscId xs = DiscID $ fromIntegral $
                length xs +
                shiftL (quot (last xs) 75) 8 +
                shiftL (rem
                        (foldl
                         (\x y -> x + sumDigits(2 + quot y 75))
                         0 (0:init xs))
                        255
                       ) 24
    where sumDigits 0 = 0
          sumDigits n = let (q,r) = quotRem n 10
                        in r + sumDigits q



CD/AccurateRip.hs
{-# LANGUAGE BangPatterns #-}

module CD.AccurateRip where

import           Data.Binary
import           Data.Binary.Get
import           Data.Binary.Put
import           Data.Bits
import           Data.List       (find)

import           CD.CDDB

type Sample = Word32

type CRC = DiscID

data ArCrcEntry = ArCrcEntry {
      id1       :: DiscID,
      id2       :: DiscID,
      cddb      :: DiscID,
      trackCRCs :: [(Word8,CRC)]  -- (confidence,crc)
    }

instance Binary ArCrcEntry where
    put (ArCrcEntry i1 i2 i3 cs) = do
                             putWord8 l
                             put i1
                             put i2
                             put i3
                             mapM_ putTrackCRCs cs
        where l = toEnum $ length cs
              putTrackCRCs (c,s) = do putWord8 c
                                      put s
                                      putWord32le 0
    get = do
      l <- getWord8
      i1 <- get
      i2 <- get
      i3 <- get
      cs <- mapM (const getTrackCRCs) [1..l]
      return $ ArCrcEntry i1 i2 i3 cs
        where getTrackCRCs = do c <- getWord8
                                s <- get
                                _ <- getWord32le
                                return (c,s)

instance Show ArCrcEntry where
    show (ArCrcEntry w1 w2 w3 xs) = "ID: " ++ show w1 ++ "-" ++
                                    show w2 ++ "-" ++
                                    show w3 ++ "\n\n" ++
                                    showEntry
        where showEntry = concat $ zipWith f [(1::Int)..] xs
              f i (c,s) | c == 0 && s == DiscID 0 = "track " ++ show i ++
                                                    ":\tnot present" ++ "\n"
                        | otherwise               = "track " ++ show i ++
                                                    ":\t" ++ show s ++
                                                    "\twith confidence\t" ++
                                                    show c ++ "\n"

showArCrcEntry :: ArCrcEntry -> String
showArCrcEntry (ArCrcEntry w1 w2 w3 xs) =
    show w1 ++ "-" ++ show w2 ++ "-" ++ show w3
                  ++ "\n" ++ "Track No.\tChecksum\tAccuracy" ++ "\n\n" ++
                  showEntry
        where showEntry = concat $ zipWith f [(1::Int)..] xs
              f i (c,s) | c == 0 && s == DiscID 0 = "Track " ++ show i ++
                                                    ":\t(not present)" ++ "\n"
                        | otherwise               = "Track " ++ show i ++
                                                    ":\t" ++ show s ++
                                                    "\t" ++ show c ++ "\n"


newtype ArData = ArData {unArData :: [ArCrcEntry]}

instance Binary ArData where
    put (ArData ents) = mapM_ put ents
    get = fmap ArData getList
        where getList = do b <- isEmpty
                           if b
                             then return []
                             else do a <- get
                                     as <- getList
                                     return (a:as)

instance Show ArData where
    showsPrec _ (ArData []) = id
    showsPrec n (ArData (x:xs)) = showsPrec n x . ('\n' :) . ('\n' :) .
                                  showsPrec n (ArData xs)

showArData :: ArData -> String
showArData (ArData xs) = concat $ zipWith f xs [1..length xs]
    where f x i = "#" ++ show i ++ " " ++ showArCrcEntry x ++ "\n"

{-
type DiscMatch = (Bool,Bool,Bool,Bool) -- track count, ID1, ID2, CDDB

data ArResult = AccuratelyRipped Word8 | NotAccuratelyRipped | NotPresent
     deriving Show

showArResult :: ArResult -> String
showArResult (AccuratelyRipped n) = "accurately ripped with confidence "
                                    ++ show n
showArResult NotAccuratelyRipped  = "not accurately ripped"
showArResult NotPresent           = "not present in the database"
-}
data RipHash = RipHash {
      ripId1  :: DiscID,
      ripId2  :: DiscID,
      ripCddb :: DiscID,
      ripCrc  :: [(CRC,CRC)]
    }

matchPressing :: ArData -> RipHash -> (Maybe [Word8], Maybe [Word8])
matchPressing (ArData ardata) hash = let cs = map trackCRCs ardata
                                     in (matchV1 cs, matchV2 cs)
    where matchV1 xs = let a = find (\ys -> snd (unzip ys) == v1Sums) xs
                       in case a of
                            Just zs -> Just $ fst $ unzip zs
                            Nothing -> Nothing
          matchV2 xs = let a = find (\ys -> snd (unzip ys) == v2Sums) xs
                       in case a of
                            Just zs -> Just $ fst $ unzip zs
                            Nothing -> Nothing
          v1Sums = fst $ unzip $ ripCrc hash
          v2Sums = snd $ unzip $ ripCrc hash

verifyTracks :: ArData -> RipHash -> [(Word,Word)]
verifyTracks (ArData ardata) (RipHash i1 i2 i3 csum) =
    zipWith (trackConf (filter p ardata)) [0..length csum - 1] csum
    where trackConf ents i = go (map (\e -> trackCRCs e !! i) ents) (0,0)
          go []         (a1,a2) _       = (a1,a2)
          go ((c,s):xs) (a1,a2) (v1,v2) =
              go xs (a1 + fromIntegral (if v1 == s then c else 0), a2 + fromIntegral (if v2 == s then c else 0)) (v1,v2)
          p (ArCrcEntry x y z _) | x == i1 && y == i2 && z == i3 = True
                                 | otherwise                     = False

{-
verifyRip :: RipHash -> ArData -> Either [ArResult] DiscMatch
verifyRip hash (ArData ardata) | ((length $ ripCrc hash) == (length $ trackCRCs (ardata!!0)),
                         ripId1 hash == id1 (ardata!!0),
                         ripId2 hash == id2 (ardata!!0),
                         ripCddb hash == cddb (ardata!!0)) /= (True,True,True,True) = Right ((length $ ripCrc hash) == (length $ trackCRCs (ardata!!0)),
             ripId1 hash == id1 (ardata!!0),
             ripId2 hash == id2 (ardata!!0),
             ripCddb hash == cddb (ardata!!0))
    | otherwise = Left (comp (ripCrc hash) (map trackCRCs ardata))
    where comp [] _ = []
          comp (x:xs) ys = (verifyTrack x (map head ys)):(comp xs (map tail ys))

verifyTrack :: CRC -> [(Word8, CRC)] -> ArResult
verifyTrack c xs | ys == [] = NotPresent
                 | zs == [] = NotAccuratelyRipped
                 | otherwise = AccuratelyRipped $ fst $ head zs
    where ys = filter (\a -> (fst a /= 0)) xs
          zs = filter (\a -> (snd a == c)) ys
-}

discId1 :: [Int] -> DiscID
discId1 = DiscID . toEnum . sum

discId2 :: [Int] -> DiscID
discId2 xs = DiscID $ toEnum $ sum $ zipWith (*) (1:xs) [1..n+1]
    where n = length xs


arUrl :: [Int] -> String
arUrl xs = "http://www.accuraterip.com/accuraterip/" ++
           (show (discId1 xs)!!7) : '/' :
           (show (discId1 xs)!!6) : '/' :
           (show (discId1 xs)!!5) : "/dBAR-" ++
           addZeros 3 (show (length xs)) ++
           '-' : show (discId1 xs) ++
           '-' : show (discId2 xs) ++
           '-' : show (cddbDiscId xs) ++
           ".bin"
    where addZeros n s = if length s >= n
                         then s
                         else addZeros n ('0':s)

{-
crcFromTo :: Int -> Int -> [Sample] -> CRC
crcFromTo m n = crc m 0
    where crc _ _ [] = 0
          crc !l !acc (y:ys) | l <= n    = crc (l+1) (acc + (toEnum l)*y) ys
                             | otherwise = acc
-}
{-
crcFromTo :: Int -> Int -> [Sample] -> CRC
crcFromTo m n = snd . foldl' step (m,0) . take (n-m+1)
    where
      step (!l,!acc) y = (l+1,acc + (toEnum l)*y)
-}


samplesPerSector :: Num a => a
samplesPerSector = 588


-- |@'droppedSamples'@ is the number of samples discarded in the
-- computation of AccurateRip crc from the beginning and the end of
-- the samples list. It is a constant equal to the length of 5 CD
-- sectors.
droppedSamples :: Num a => a
droppedSamples = 5 * samplesPerSector


-- |@'listCrcs'@ computes the AccurateRip crc from each track.
listCrcs :: Int       -- ^ Offset in samples
         -> [Int]     -- ^ List of track lengths in samples
         -> [Sample]  -- ^ List of samples extracted from the audio CD
         -> [CRC]     -- ^ Returns the list of CRC's corresponding to
                     -- each track
{-listCrcs _ [] _ = []
listCrcs off [x] dat = [crcFromTo droppedSamples
                                  (x - droppedSamples)
                                  (drop (droppedSamples - 1 + off) dat)]
listCrcs off (x:xs) dat = a `seq` (a : (reverse $ go (drop (x+off) dat) xs))
    where a = crcFromTo droppedSamples
              x
              (drop (droppedSamples - 1 + off) dat)
          go _ [] = []
          go d [y] = [crcFromTo 1 (y-droppedSamples) d]
          go d (y:ys) = b `seq` (b : go (drop y d) ys)
                 where b = crcFromTo 1 y d-}
listCrcs = generalListCrcs crc

listCrcsv2 :: Int -> [Int] -> [Sample] -> [CRC]
listCrcsv2 = generalListCrcs crcv2

generalListCrcs :: ([Sample] -> CRC) -> Int -> [Int] -> [Sample] -> [CRC]
generalListCrcs f off ls = map f . chunks ls . zeroEnd (sum ls) .
                           zeroInit . applyOff off

chunks :: [Int] -> [Sample] -> [[Sample]]
chunks [] _ = []
chunks (l:ls) xs = case splitAt l xs of
                     (xs1,xs2) -> xs1 : chunks ls xs2

dropFirstLast :: Int -> [[Sample]] -> [[Sample]]
dropFirstLast _ [] = []
dropFirstLast l [x] = [drop (droppedSamples - 1) $ take (l - droppedSamples) x]
dropFirstLast l (x:xs) = drop (droppedSamples - 1) x : go xs
    where go [] = []
          go [y] = [take (l - droppedSamples) y]
          go (y:ys) = y : go ys

zeroInit :: [Sample] -> [Sample]
zeroInit =
    zipWith (\i x -> if i < droppedSamples - 1 then 0 else x) [(0::Int)..]

zeroEnd :: Int -> [Sample] -> [Sample]
zeroEnd l = zipWith (\i x -> if i >= l - droppedSamples then 0 else x) [0..]

applyOff :: Int -> [Sample] -> [Sample]
applyOff off dat | off >= 0 = drop off dat ++ replicate off 0
                 | otherwise = replicate off 0 ++ dat

crc :: [Sample] -> CRC
--crc = foldl1' (+) . zipWith (*) [1..]
crc = DiscID . go 0 1
    where go acc _ [] = acc
          go !acc !n (x:xs) = go (acc + n*x) (n+1) xs

crcv2 :: [Sample] -> CRC
crcv2 = DiscID . go 0 1
    where go acc _ [] = acc
          go !acc !n (x:xs) = go (acc + f n x) (n+1) xs
          f :: Word64 -> Sample -> Word32
          f n x = let a = n * fromIntegral x
                  in fromIntegral (a .&. 0xFFFFFFFF) +
                     fromIntegral (shiftR (a .&. 0xFFFFFFFF00000000) 32)

pairCrc :: [Sample] -> (CRC,CRC)
pairCrc = go (0,0) 1
    where go (a1,a2)  _  [] = (DiscID a1, DiscID a2)
          go (!a1,!a2) !n (x:xs) = go (a1 + n*x, a2 + f n x) (n+1) xs
          f n x = let a = fromIntegral n * fromIntegral x :: Word64
                  in fromIntegral (a .&. 0xFFFFFFFF) +
                     fromIntegral (shiftR (a .&. 0xFFFFFFFF00000000) 32)

listPairCrcs :: Int -> [Int] -> [Sample] -> [(CRC,CRC)]
listPairCrcs off ls = map pairCrc . chunks ls . zeroEnd (sum ls) .
                      zeroInit . applyOff off


Main.hs
import           Control.DeepSeq
import           Control.Monad
import           Data.Binary
import           Data.Bits
import           Data.String
import           Data.Version
import           Network.HTTP
import           Options.Applicative
import           System.Environment
import           System.Exit
import           System.IO
import           System.Process

import           CD.AccurateRip
import           CD.CDDB

import           Paths_flaccuraterip

progName :: String
progName = "flAccurateRip"

intro :: String
intro = progName ++ " " ++ showVersion version ++
        "\n\
        \Copyright (C) 2012-2016 Nicola Squartini.\n\
        \License GPLv3+: GNU GPL version 3 or \
        \later <http://gnu.org/licenses/gpl.html>\n\
        \This is free software: you are free to change and redistribute it.\n\
        \There is NO WARRANTY, to the extent permitted by law."

bugReport :: String
bugReport = "Report bugs to: <https://github.com/tensor5/flAccurateRip/issues>"

data Options = Options
    { opt30Samples :: Bool
    , optOffset    :: Int
    , optShowEntry :: Bool
    , optVerbose   :: Bool
    , optFiles     :: [String]
    }

options :: Parser Options
options = Options
          <$> switch (short 'c'
                      <> long "with-30-samples-correction"
                      <> help "Add 30 samples to the offset (use if the CD was \
                              \ripped using the correct offset, which is 30 \
                              \samples less than that in \
                              \<http://www.accuraterip.com/driveoffsets.htm>)"
                     )
          <*> option auto (short 'o'
                           <> long "with-sample-offset"
                           <> metavar "N"
                           <> value 0
                           <> help "Set ripping offset to N"
                          )
          <*> switch (short 's'
                      <> long "show-database-entry"
                      <> help "Show the AccurateRip database entry for this rip"
                     )
          <*> switch (short 'v'
                      <> long "verbose"
                      <> help "Be verbose"
                     )
          <*> some (argument str $
                    metavar "track01.flac track02.flac ... trackNN.flac"
                   )

versionOption :: Parser (a -> a)
versionOption = infoOption intro (long "version"
                                  <> help "Show version information and exit")

lengthsToOffsets :: Num a => [a] -> [a]
lengthsToOffsets = scanl1 (+)

offsetsToLengths :: Num a => [a] -> [a]
offsetsToLengths [] = []
offsetsToLengths [x] = [x]
offsetsToLengths xs = offsetsToLengths i ++ [last xs - last i]
    where i = init xs


getFlacSampleNumber :: FilePath -> IO Int
getFlacSampleNumber file = do
  (_,b, _) <- readProcessWithExitCode
              "metaflac"
              ["--show-total-samples", file]
              ""
  return (read b)


getOffsetsFromFlacs :: [FilePath] -> IO [Int]
getOffsetsFromFlacs [file] = do
  (a,_,_) <- readProcessWithExitCode
             "metaflac"
             ["--export-cuesheet-to=-", file]
             ""
  if a == ExitSuccess
    then do
      prog <- getProgName
      putStrLn (prog ++ ": sorry, rips to single file are not supported yet")
      exitFailure
    else do
      l <- getFlacSampleNumber file
      return [quot l samplesPerSector]
getOffsetsFromFlacs filelist = do
  s <- mapM getFlacSampleNumber filelist
  return $ lengthsToOffsets (map (`quot` samplesPerSector) s)




stringToWord32List :: String -> [Word32]
stringToWord32List [] = []
stringToWord32List (x1:x2:x3:x4:xs) =
    fourCharToWord32 x1 x2 x3 x4 : stringToWord32List xs
  where fourCharToWord32 c1 c2 c3 c4 =
            fromIntegral (fromEnum c1) +
            fromIntegral (fromEnum c2) `shiftL` 8 +
            fromIntegral (fromEnum c3) `shiftL` 16 +
            fromIntegral (fromEnum c4) `shiftL` 24
stringToWord32List _ = error "stringToWord32List: length of String should be \
                             \divisible by 4."


pipeFlacs :: [FilePath] -> IO Handle
pipeFlacs xs = do
  (_,Just b,_,_) <- createProcess (CreateProcess
                                   (RawCommand "flac" ("-s":"-c":"-d":
                                                       "--force-raw-format":
                                                       "--endian=little":
                                                       "--sign=signed":xs))
                                   Nothing
                                   Nothing
                                   Inherit
                                   CreatePipe
                                   Inherit
                                   True
                                   True
                                   False
                                   False
                                   False
                                   False
                                   Nothing
                                   Nothing
                                   False)
  return b

showAccuracy :: ArData -> RipHash -> IO ()
showAccuracy ar hash = showRes (verifyTracks ar hash) (ripCrc hash)

showRes :: [(Word,Word)] -> [(CRC,CRC)] -> IO ()
showRes x y = do
  putStrLn "Track No.\tAccuracy (v.1 sum)\tAccuracy (v.2 sum)\n"
  sR (1::Int) x y
    where sR _ [] _ = return ()
          sR _ _ [] = return ()
          sR i ((c1,c2):cs) ((s1,s2):ss) = do
            putStrLn ("Track " ++ show i ++ ":\t" ++
                      show c1 ++ " (" ++ show s1 ++ ")\t\t" ++
                      show c2 ++ " (" ++ show s2 ++ ")")
            sR (i+1) cs ss

main :: IO ()
main = do
  opts <- execParser $ info (helper <*> versionOption <*> options)
          (fullDesc
           <> progDesc "Verify the given FLAC files"
           <> footer bugReport)
  let flaclist = optFiles opts
      verbose  = optVerbose opts
  list <- getOffsetsFromFlacs flaclist
  putStr "Connecting to AccurateRip database"
  when verbose $ putStr (" (<"  ++ arUrl list ++ ">)")
  putStr "..."
  hFlush stdout
  res <- simpleHTTP (getRequest $ arUrl list)
  case res of
    Right (Response (4,0,4) _ _ _) -> do
             putStrLn " no pressing found."
             exitFailure
    _ -> do
         ardatastr <- getResponseBody res
         putStrLn " OK"
         let ardata = decode $ fromString ardatastr
         when (optShowEntry opts) $ do
             let n = length $ unArData ardata
             if n == 1
               then do putStrLn "Found 1 pressing:\n"
                       putStrLn $ showArData ardata
               else do putStrLn ("Found " ++ show n ++ " pressings:\n")
                       putStrLn $ showArData ardata
         let offset = optOffset opts + 30 * fromEnum (opt30Samples opts)
         putStr "Computing track checksums from input files"
         when verbose $ putStr (" (using " ++ show offset ++ " samples offset)")
         putStrLn "..."
         a <- pipeFlacs flaclist
         hSetBinaryMode a True
         fl <- hGetContents a
         let cs = listPairCrcs offset
                  (map (\n -> samplesPerSector * n) (offsetsToLengths list))
                  (stringToWord32List fl)
         (showAccuracy ardata .
          RipHash (discId1 list) (discId2 list) (cddbDiscId list))
           $!! cs
         hClose a
